README.md
# Changelog

![Licence](https://img.shields.io/badge/license-MIT-green.svg?style=for-the-badge)

A command line utility to create changelogs. A variation of the [changelog utility](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/bin/changelog) used in [gitlab](https://gitlab.com). Allows for the 

## Pre-requisites

- [golang](https://golang.org/) 1.11+
- [tusk](https://github.com/rliebz/tusk)
- [dep](https://github.com/golang/dep)

## Conventions
- **`Changelog.md`** must reside in root folder and must have top level header **# Changelog**
- Must run command in same folder as **`Changelog.md`**

## Workflow

1. Create a new feature on a branch
2. Create a new changelog entry for feature 
    - Creates a folder called `unreleased` in root folder
    - Multiple people can be on their own branches with their own entries
3. Merge work into release branch when ready
4. Preview the changelog `changelog release vX.X.X --preview`
5. Update the changelog `changelog release vX.X.X` 
    - Version string can follow your preferred format
    - Will create a file called `changelog.yml` with all changes in root folder
6. Commit the changes 
    - Unreleased folder will be deleted
    - Can optionally keep the yaml files
7. Cake

## Development

Once you are satisfied with the current version of the code run the following:

```shell
tusk test
tusk release --version=X.X.X
git add -u
git tag vX.X.X
git commit -m"Latest release"
```

This performs the following:
- Merge all elements into `changelog.yml`
- Add all elements to `CHANGELOG.md`
- copy the latest version to `tools` folder
- Add all the updated file (includes the version in source code and the tool itself)
- Tag the latest commit
- Submit changes

CHANGELOG.md
# Changelog

## v0.1.0

**NEW** create changelog entry  
**NEW** create final changelog  
**NEW** display version information  


cmd/add_test.go
package cmd_test

import (
	"os"
	"testing"

	"github.com/spf13/cobra"
	"github.com/stretchr/testify/assert"

	"github.com/acco32/changelog/changelog"
	"github.com/acco32/changelog/cmd"
)

func TestErrorWhenNoArgs(t *testing.T) {
	addRootCmd := &cobra.Command{Use: "changelog"}
	addRootCmd.AddCommand(cmd.AddCmd)

	_, out, err := executeCommandC(addRootCmd, "add")
	assert.Contains(t, out, "Error")
	assert.NotNil(t, err)
}

func init() {
	os.RemoveAll(changelog.DefaultChangelogFolder)
}

func TestNoTitleError(t *testing.T) {
	addRootCmd := &cobra.Command{Use: "changelog"}
	addRootCmd.AddCommand(cmd.AddCmd)

	_, out, err := executeCommandC(addRootCmd, "add", changelog.Fixed.Name, "-a author")
	assert.Contains(t, out, "Error")
	assert.Contains(t, out, "\"title\" not set")
	assert.NotNil(t, err)
}

func TestNoAuthorError(t *testing.T) {
	t.Skip("Test causing goroutine error. Need to investigate.")
	addRootCmd := &cobra.Command{Use: "changelog"}
	addRootCmd.AddCommand(cmd.AddCmd)

	_, out, err := executeCommandC(addRootCmd, "add", changelog.Fixed.Name, "-t title")
	assert.Contains(t, out, "Error")
	assert.Contains(t, out, "\"author\" not set")
	assert.NotNil(t, err)
}

func TestErrorWithUnknownType(t *testing.T) {
	addRootCmd := &cobra.Command{Use: "changelog", Args: cobra.ExactArgs(1)}
	addRootCmd.AddCommand(cmd.AddCmd)

	_, out, err1 := executeCommandC(addRootCmd, "add", "blah", "-t title", "-a author")
	assert.Contains(t, out, "Unrecognized format entered")
	assert.Contains(t, out, "Unrecognized format entered")
	assert.Nil(t, err1)

	_, err2 := os.Stat(changelog.DefaultChangelogFolder)
	assert.True(t, os.IsNotExist(err2))
}


cmd/cmd_test.go
package cmd_test

import (
	"bytes"

	"github.com/spf13/cobra"
)

func executeCommandC(root *cobra.Command, args ...string) (c *cobra.Command, output string, err error) {
	buf := new(bytes.Buffer)
	root.SetOutput(buf)
	root.SetArgs(args)

	c, err = root.ExecuteC()

	return c, buf.String(), err
}


cmd/info_test.go
package cmd_test

import (
	"testing"

	"github.com/spf13/cobra"
	"github.com/stretchr/testify/assert"

	"github.com/acco32/changelog/cmd"
)

func TestSummaryDisplaysCopyright(t *testing.T) {
	infoRootCmd := &cobra.Command{Use: "changelog"}
	infoRootCmd.AddCommand(cmd.InfoCmd)

	_, out, err := executeCommandC(infoRootCmd, "info")
	assert.Contains(t, out, "© 2018")
	assert.Nil(t, err)
}

func TestErrorWithInvalidArgNumber(t *testing.T) {
	infoRootCmd := &cobra.Command{Use: "changelog"}
	infoRootCmd.AddCommand(cmd.InfoCmd)

	_, out, err := executeCommandC(infoRootCmd, "info", "blah")
	assert.Contains(t, out, "Error")
	assert.NotNil(t, err)
}


cmd/add.go
package cmd

import (
	"fmt"

	"github.com/acco32/changelog/changelog"
	"github.com/spf13/cobra"
)

var author string
var changelogType changelog.Changelog
var title string

func init() {
	AddCmd.Flags().StringVarP(&author, "author", "a", "", "Individual creating feature. Limited to 80 characters.")
	AddCmd.MarkFlagRequired("author")
	AddCmd.Flags().StringVarP(&title, "title", "t", "", "Description. Space will be replaced by underscores. Limited to 120 characters.")
	AddCmd.MarkFlagRequired("title")
}

var AddCmd = &cobra.Command{
	Use:   "add [TYPE]",
	Short: "Add new changelog TYPE",
	Long: fmt.Sprintf("Add new changelog TYPE from:\n\n%s\n%s\n%s\n%s\n%s\n%s\n%s",
		changelog.New.ToString(),
		changelog.Fixed.ToString(),
		changelog.Changed.ToString(),
		changelog.Deprecated.ToString(),
		changelog.Performance.ToString(),
		changelog.Security.ToString(),
		changelog.Removed.ToString()),
	Args: cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		if err := add(args); err != nil {
			displayErrorAndExit(cmd, err)
		}
	},
}

func add(args []string) error {

	parseEntryType := func(changelogType string) (changelog.Changelog, error) {

		switch changelogType {
		case changelog.New.Name:
			return changelog.New, nil
		case changelog.Fixed.Name:
			return changelog.Fixed, nil
		case changelog.Changed.Name:
			return changelog.Changed, nil
		case changelog.Removed.Name:
			return changelog.Removed, nil
		case changelog.Deprecated.Name:
			return changelog.Deprecated, nil
		case changelog.Security.Name:
			return changelog.Security, nil
		case changelog.Performance.Name:
			return changelog.Performance, nil
		}
		return changelog.Other, fmt.Errorf("Unrecognized format entered: %s", changelogType)
	}

	entryType, err := parseEntryType(args[0])
	if err != nil {
		return err
	}

	entry := changelog.Entry{Title: title, Author: author, Type: entryType}
	return changelog.CreateChangelogEntry(entry, changelog.DefaultChangelogFolder)
}


cmd/release.go
package cmd

import (
	"fmt"
	"os"

	"github.com/acco32/changelog/changelog"
	"github.com/spf13/cobra"
)

var preview bool

func init() {
	ReleaseCmd.Flags().BoolVarP(&preview, "preview", "p", false, "Show preview of changelog in terminal. Do not delete existing unreleased files.")
}

var ReleaseCmd = &cobra.Command{
	Use:   "release [VERSION]",
	Short: "Generate Changelog with version string VERSION",
	Args:  cobra.ExactArgs(1),
	Long:  "Use data template and generate/append latest changes to changelog. This will delete data entries in unreleased folder.",
	Run: func(cmd *cobra.Command, args []string) {
		if err := release(args, changelog.DefaultChangelogFile); err != nil {
			displayErrorAndExit(cmd, err)
		}
	},
}

func release(args []string, changelogFile string) error {


	err := changelog.CreateChangelog(changelog.DefaultChangelogFolder, changelog.DefaultChangelogDataFile)
	if err != nil {
		return err
	}

	_, err = os.Stat(changelog.DefaultChangelogDataFile)
	if err != nil {
		return fmt.Errorf("%s not found", changelog.DefaultChangelogDataFile)
	}

	if preview {
		t, e := changelog.Text(args[0], changelog.DefaultChangelogDataFile)
		if e != nil {
			return e
		}
		fmt.Println(t)
		os.Remove(changelog.DefaultChangelogDataFile)
		return nil
	}

  md, err := changelog.Markdown(args[0], changelog.DefaultChangelogDataFile)
  if err != nil {
    return err
  }

  err = changelog.AppendChanges(changelog.DefaultChangelogFile, md)
  if err != nil {
    return err
  }

	return os.RemoveAll(changelog.DefaultChangelogFolder)
}


changelog/changelog.go
package changelog

import (
	"bytes"
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"strings"

	"github.com/smallfish/simpleyaml"
)

//Changelog Template for type of change being applied
type Changelog struct {
	Name        string
	Description string
}

// ToString Prints friendly description
func (c *Changelog) ToString() string {
	return fmt.Sprintf("%12s - %s", c.Name, c.Description)
}

//Entry A changelog entry
type Entry struct {
	Title  string
	Author string
	Type   Changelog
}

var (
	//New New Feature
	New = Changelog{Name: "new", Description: "New feature"}
	//Fixed Bug Fix
	Fixed = Changelog{Name: "fixed", Description: "Bug fix"}
	//Changed Feature change
	Changed = Changelog{Name: "changed", Description: "Feature change"}
	//Deprecated New deprecation
	Deprecated = Changelog{Name: "deprecated", Description: "Feature deprecation"}
	//Removed Feature removal
	Removed = Changelog{Name: "removed", Description: "Feature removal"}
	//Security Security Fix
	Security = Changelog{Name: "security", Description: "Security fix"}
	//Performance Performance Improvement
	Performance = Changelog{Name: "performance", Description: "Performance improvement"}
	//Other Unknown Changelog type
	Other = Changelog{Name: "other", Description: "Other"}
)

const (
	//DefaultChangelogFolder All entries are stored in an unreleased folder.
	DefaultChangelogFolder = "unreleased"

	// DefaultChangelogFile The name of the file to append rendered content
	DefaultChangelogFile = "CHANGELOG.md"

	// DefaultChangelogDataFile The name of the file to append unreleased data
	DefaultChangelogDataFile = "changelog.yml"

	// DefaultChangelogHeader The header text to seek which is where we append changes
	DefaultChangelogHeader = "# Changelog"
)

//CreateChangelogEntry Create a new entry for the changelog. Usually done on a new branch.
func CreateChangelogEntry(file Entry, unreleasedFolder string) error {

	if !validTitle(file.Title) {
		return fmt.Errorf("Title must be between 1 and 120 characters (including spaces). Current Length %d", len(file.Title))
	}

	if !validAuthor(file.Author) {
		return fmt.Errorf("Author must be between 1 and 80 characters. Current Length %d", len(file.Author))
	}

	filename := path.Join(unreleasedFolder, strings.ToLower(strings.Replace(file.Title, " ", "_", -1))+".yml")
	if fileExists(filename) {
		return errors.New("File already exists")
	}

	os.Mkdir(unreleasedFolder, 0777)

	f, err := os.Create(filename)
	if err != nil {
		return err
	}

	yaml := fmt.Sprintf("-\n  title: %s\n  author: %s\n  type: %s\n", file.Title, file.Author, file.Type.Name)

	if _, err := f.WriteString(yaml); err != nil {
		return err
	}

	if err := f.Close(); err != nil {
		return err
	}

	return nil
}

func fileExists(filename string) bool {
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		return false
	}
	return true
}

func validTitle(title string) bool {
	return len(title) > 0 && len(title) < 120
}

func validAuthor(author string) bool {
	return len(author) > 0 && len(author) < 80
}

// CreateChangelog Gather all unreleased entries and create latest log
func CreateChangelog(unreleasedFolder string, outputFilename string) error {

	_, err := os.Stat(unreleasedFolder)
	if err != nil {
		return fmt.Errorf("\"%s\" folder does not exist", unreleasedFolder)
	}

	files, err := ioutil.ReadDir(unreleasedFolder)
	if err != nil {
		return fmt.Errorf("Unable to read files in unreleased folder: %s", err.Error())
	}

	if len(files) == 0 {
		return fmt.Errorf("At least one file must exist")
	}

	var buffer bytes.Buffer

	for _, f := range files {
		dat, err := ioutil.ReadFile(path.Join(unreleasedFolder, f.Name()))
		if err != nil {
			return fmt.Errorf("Error reading file %s", f.Name())
		}

		buffer.WriteString(string(dat))
	}

	cl, _ := os.Create(outputFilename)
	defer cl.Close()

	cl.Write(buffer.Bytes())

	return nil
}

// Text Generate a text representation of latest changes
func Text(version string, outputFilename string) (string, error) {

	if version == "" {
		return "", errors.New("version string cannot be empty")
	}

	data, err := ioutil.ReadFile(outputFilename)
	if err != nil {
		return "", err
	}

	y, err := simpleyaml.NewYaml(data)
	if err != nil {
		return "", fmt.Errorf("Problem reading data file: %s", err.Error())
	}

	if !y.IsArray() {
		return "", fmt.Errorf("input yaml has incorrect schema layout. Expected only an list of items")
	}

	text := fmt.Sprintf("%s:\n", version)
	totalElements, _ := y.GetArraySize()
	for cl := 0; cl < totalElements; cl++ {
		featureTitle, _ := y.GetIndex(cl).Get("title").String()
		featureType, _ := y.GetIndex(cl).Get("type").String()
		text = fmt.Sprintf("%s%12s | %s\n", text, featureType, featureTitle)
	}

	return text, nil
}

// Markdown Generate markdown representation of latest changes
func Markdown(version string, dataFilename string) (string, error) {

	if version == "" {
		return "", errors.New("version string cannot be empty")
	}

	data, err := ioutil.ReadFile(dataFilename)
	if err != nil {
		return "", err
	}

	y, err := simpleyaml.NewYaml(data)
	if err != nil {
		return "", fmt.Errorf("Problem reading data file: %s", err.Error())
	}

	if !y.IsArray() {
		return "", fmt.Errorf("input yaml has incorrect schema layout. Expected only an list of items")
	}

	text := fmt.Sprintf("## %s\n\n", version)
	totalElements, _ := y.GetArraySize()
	for cl := 0; cl < totalElements; cl++ {
		featureTitle, _ := y.GetIndex(cl).Get("title").String()
		featureType, _ := y.GetIndex(cl).Get("type").String()
		text = fmt.Sprintf("%s**%s**  %s  \n", text, strings.ToUpper(featureType), featureTitle)
	}
	return text, nil
}

// AppendChanges Add the latest changes to current changelog file
func AppendChanges(changelogFile string, changes string) error {
	_, err := os.Stat(changelogFile)
	if err != nil {
		return fmt.Errorf("%s not found", changelogFile)
	}

	clf, err := ioutil.ReadFile(changelogFile)
	if err != nil {
		return fmt.Errorf("Problem reading %s file", changelogFile)
	}

	text := string(clf)
	if !strings.Contains(text, DefaultChangelogHeader) {
		return fmt.Errorf("Missing Changelog Header, \"%s\", in %s file", DefaultChangelogHeader, changelogFile)
	}

	newChanges := strings.Replace(text, DefaultChangelogHeader, fmt.Sprintf("%s\n\n%s", DefaultChangelogHeader, changes), 1)
	return ioutil.WriteFile(changelogFile, []byte(newChanges), 0644)
}
